src/commands/signzone.rs

```diff
@@ -1418,41 +1419,67 @@ impl SignZone {
         Ok(records)
     }
 
-    fn find_apex(
-        records: &SortedRecords<StoredName, StoredRecordData, MultiThreadedSorter>,
-    ) -> Result<(StoredName, Class, Ttl, Serial), Error> {
-        let soa = match records.find_soa() {
-            Some(soa) => soa,
-            None => {
-                return Err(Error::from("Invalid zone file: Cannot find SOA record"));
+    // TODO: Add tests.
+    fn find_apex<'a>(
+        records: &'a SortedRecords<StoredName, StoredRecordData, MultiThreadedSorter>,
+        origin: Option<&StoredName>,
+    ) -> Result<&'a Record<StoredName, StoredRecordData>, Error> {
+        // If an expected origin was supplied, the found SOA must match it.
+        if let Some(expected_origin) = origin {
+            return records
+                .iter()
+                .find(|rr| rr.rtype() == Rtype::SOA && rr.owner() == expected_origin)
+                .ok_or(format!("SOA RR not found for origin '{expected_origin}'.").into());
+        }
+
+        if records.is_empty() {
+            return Err("Invalid zone file: No records found".into());
+        }
+
+        // SAFETY: We just verified that records is not empty.
+        let mut iter = records.iter().peekable();
+        let mut zone_name = iter.peek().unwrap().owner();
+        while let Some(rr) = iter.next() {
+            // Only match SOA RRs at the apex of a zone. If the current name
+            // is a child of the last seen name skip it because as the records
+            // are in sorted order children appear in the iteration order
+            // after their parent.
+            if rr.owner() != zone_name {
+                if rr.owner().starts_with(zone_name) {
```

Comment:
I guess this should be ends_with?

Reply:
That somewhat proves the point about testing as the coverage report showed this condition was not tested.